scrape/scrape.go

```diff
@@ -133,22 +133,96 @@ var (
 			Help: "How many times a scrape cache was flushed due to getting big while scrapes are failing.",
 		},
 	)
+	targetMetadataCache = newMetadataMetricsCollector()
 )
 
+// MetadataMetricsCollector is a Custom Collector for the metadata cache metrics.
+type MetadataMetricsCollector struct {
+	CacheEntries *prometheus.Desc
+	CacheBytes   *prometheus.Desc
+
+	scrapeManager *Manager
+
+	mtx sync.Mutex
+}
+
+func newMetadataMetricsCollector() *MetadataMetricsCollector {
+	return &MetadataMetricsCollector{
+		CacheEntries: prometheus.NewDesc(
+			"prometheus_target_metadata_cache_entries",
+			"Total number of metric metadata entries in the cache",
+			[]string{"scrape_job"},
+			nil,
+		),
+		CacheBytes: prometheus.NewDesc(
+			"prometheus_target_metadata_cache_bytes",
+			"The number of bytes that are currently used for storing metric metadata in the cache",
+			[]string{"scrape_job"},
+			nil,
+		),
+	}
+}
+
+func (mc *MetadataMetricsCollector) registerManager(m *Manager) {
+	mc.scrapeManager = m
+}
+
+// Describe sends the metrics descriptions to the channel.
+func (mc *MetadataMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- mc.CacheEntries
+	ch <- mc.CacheBytes
+}
+
+// Collect creates and sends the metrics for the metadata cache.
+func (mc *MetadataMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	mc.mtx.Lock()
+	defer mc.mtx.Unlock()
+
+	if mc.scrapeManager == nil {
+		return
+	}
+
+	for _, p := range mc.scrapeManager.scrapePools {
```

Comment:
This needs locking. It'd be better to use TargetsActive.

Reply:
I was hoping to avoid using `TargetsActive` as I don't need / can't use the serialization into a `Target`.  Furthermore, `Target` interacts with the `scrapeCache` for metadata via an interface which would also require modification.

On second thoughts, I think it's better to use the manager directly for this information. I can mimic most of what `TargetActive` is doing and thus limit the amount of scrape locking. 

Let me give that a try.